Applied-Upstream: commit: 5a799500f95dd207662a8dacd422d56c535392ef
Author: Evan Maddock <5157277+EbonJaeger@users.noreply.github.com>
Last-Update: 2022-03-19
Description: [PATCH] Re-connect to the Raven signal to clear all notifications
 (#76)

Fixes #75

Signed-off-by: Evan Maddock <maddock.evan@vivaldi.net>
---
 src/raven/notifications_view.vala | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/src/raven/notifications_view.vala b/src/raven/notifications_view.vala
index f61f35d9..65a09123 100644
--- a/src/raven/notifications_view.vala
+++ b/src/raven/notifications_view.vala
@@ -63,6 +63,11 @@ namespace Budgie {
 		public abstract void toggle_do_not_disturb() throws DBusError, IOError;
 	}
 
+	[DBus (name="org.budgie_desktop.Raven")]
+	public interface RavenInterface : Object {
+		public signal void ClearAllNotifications();
+	}
+
 	public class NotificationsView : Gtk.Box {
 		private const string BUDGIE_PANEL_SCHEMA = "com.solus-project.budgie-panel";
 		private const string APPLICATION_SCHEMA = "org.gnome.desktop.notifications.application";
@@ -83,6 +88,8 @@ namespace Budgie {
 		private HashTable<string, NotificationGroup> notification_groups { private get; private set; default = null; }
 		private Settings budgie_settings { private get; private set; default = null; }
 
+		private RavenInterface raven { get; private set; default = null; }
+
 		construct {
 			this.budgie_settings = new Settings(BUDGIE_PANEL_SCHEMA);
 
@@ -140,6 +147,15 @@ namespace Budgie {
 				null,
 				on_dbus_get
 			);
+
+			Bus.get_proxy.begin<RavenInterface>(
+				BusType.SESSION,
+				RAVEN_DBUS_NAME,
+				RAVEN_DBUS_OBJECT_PATH,
+				0,
+				null,
+				on_raven_get
+			);
 		}
 
 		private void on_dbus_get(Object? o, AsyncResult? res) {
@@ -155,6 +171,15 @@ namespace Budgie {
 			}
 		}
 
+		private void on_raven_get(Object? o, AsyncResult? res) {
+			try {
+				this.raven = Bus.get_proxy.end(res);
+				this.raven.ClearAllNotifications.connect(this.clear_all);
+			} catch (Error e) {
+				critical("Unable to connect to Raven via DBUS: %s", e.message);
+			}
+		}
+
 		private void on_notification_added(
 			string app_name,
 			uint32 id,
-- 
2.34.1

